### PR TITLE
Ignore cache when eslint rules have changed

### LINKED ISSUE #124: Cache persists when eslint rules have changed
The eslint-loader cache is only ever invalidated when a file is changed. This means that if you make changes to your eslint rules, the last cached lint result will still be output by eslint-loader and not the outcome based on the updated rules. The only way to fix this is to either modify each file being linted or manually delete the cache.

The entire cache should be invalidated if the eslint rules have changed.

### PR DESCRIPTION
**Pull request by wrakky**
This is an attempt to fix #124

I've tried the following approach to tackle the problem:
- Get a lint rules object for each CLIEngine instance (using [CLIEngine.getConfigForFile](http://eslint.org/docs/developer-guide/nodejs-api#getconfigforfile))
  - This is taken from the rules of the first file to be processed for each engine (I assume every file processed by the same engine will share the same lint rules)
- Hash the rules object
- Include the hashed rule in the cache against each file so it can also be compared when deciding whether to use the cache

This seems to function well in my testing and changing the lint rules ignores previously cached results. I did try using a rules hash per file instead of per engine but it was not performant at all - each time it fetches the rules for a file takes between 70-90ms which adds significant overhead for a large project (I saw 15 seconds added to my build containing 350 files when using this method whereas my submitted solution has minimal time increases)

The only caveat I can think of is that if you change the eslint rules whilst the webpack build is in watch mode and then change a file, the rules hash will not be updated in the cache so when the build is stopped and started again the file will not be cached because it will have a new rules hash to compare again but I don't think this is anything to be concerned about.

~~I still have tests to add and will do so if you are happy with this solution.~~ I have added a test for the cache option which also covers rule caching.

### CODE REVIEW

File: test/cache.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+var fs = require("fs")
+
+var cacheFilePath = "./node_modules/.cache/eslint-loader/data.json"
+
+test("eslint-loader can cache results", function(t) {
+
+  // delete the require cache for eslint-loader otherwise any previously run
+  // tests will have initialised the cache as false and prevent this test
+  // from creating the cache file
+  delete require.cache[require.resolve("../index.js")]
```
Comment (wrakky) [inline test/cache.js]: I don't know if this is the best solution but all tests share the same instance of eslint-loader so without clearing the require cache, the cache option get initialised as false in previous tests and then it is impossible set it again

Comment (MoOx) [inline test/cache.js]: Maybe we should use another test runner that will use different processes?

Comment (MoOx) [inline test/cache.js]: Maybe we should use jest or ava to get separate process?

Comment (wrakky) [inline test/cache.js]: That sounds sensible but out of scope for this PR. I'll create a new issue to gather thoughts on a restructure to testing and then this PR can wait for that.

**Comment by MoOx** [comment]
I guess you can rebase on master and now benefit of the new test architecture :)

**Comment by wrakky** [comment]
I appear to have messed up rebasing somehow so have created a new PR for this

**Pull request closed without merging**